clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -824,6 +830,29 @@ private void handshakeFailure(SSLException sslException, boolean flush) throws I
             throw handshakeException;
     }
 
+    // SSL handshake failures are typically thrown as SSLHandshakeException, SSLProtocolException,
+    // SSLPeerUnverifiedException or SSLKeyException if the cause is known. But the SSL engine may
+    // throw exceptions using the base class SSLException in a few cases:
+    //   a) If there are no matching ciphers or TLS version or the private key is invalid, client will be
+    //      unable to process the server message and an SSLException is thrown:
+    //      javax.net.ssl.SSLException: Unrecognized SSL message, plaintext connection?
+    //   b) If server closes the connection gracefully during handshake, client may receive close_notify
+    //      and and an SSLException is thrown:
+    //      javax.net.ssl.SSLException: Received close_notify during handshake
+    // We want to handle a) as a non-retriable SslAuthenticationException and b) as a retriable IOException.
+    // To do this we need to rely on the exception string. Since it is safer to throw a retriable exception
+    // when we are not sure, we will treat only the first exception string as a handshake exception.
+    private void maybeProcessHandshakeFailure(SSLException sslException, boolean flush, IOException ioException) throws IOException {
```

Comment:
Would it be easier to understand if this handled all of the unwrap exceptions after the IOException? And then we could call this method `processUnwrapExceptionAfterIOException`.

Reply:
@ijuma Thanks for the review. At the moment, this method is used to process `SSLException` in two places, regardless of whether the original exception was after `IOException` or not.  I thought it would be better to do the String check in a single method rather than separate out handling of `IOException`. We need to handle both cases because `SSLException` due to `close_notify` may be processed before or after `IOException`. I can use two methods if a single method is confusing.

Reply:
@ijuma Yes, I was in two minds about that - whether to use `instanceof` in one place or catch the exception in two places. I have updated to use the common method. Thanks.